Grafica series de tiempo

efren-cabrera opened the pull request:
En esta rama se hicieron muchos cambios, se separó la aplicación en tres partes. Cada una corre en un contenedor y para que todas funcionen juntas se utiliza `docker-compose`.

Una parte corresponde al _front-end_, otra se encarga de proporcionar una api para poder obenter los datos y la última se encarga de calcular los valores de lambda a partir de los datos que le pasa el _front-end_.

Aunque parezca que se modificacon muchos archivos, en realidad la mayoría solo se cambiaron de lugar.

efren-cabrera committed:
Merge branch 'feature/add_data' into feature/agrega_api_lambdas

efren-cabrera committed:
Cambia dirección de archivos

efren-cabrera committed:
Mueve front a otra carpeta para separar los servicios

efren-cabrera committed:
Separa la aplicación en dos servicios

efren-cabrera committed:
Cambia entrypoints

efren-cabrera committed:
Agrega header que permitir peticiones desde otros servicio

efren-cabrera committed:
Agrega primer estructura del folder que contiene la api que calcula las lambda

efren-cabrera committed:
Agrega archivo de configuración para utilizar docker-compose

efren-cabrera committed:
Agrega primer versión de la aplicación para calcular las lambda

efren-cabrera committed:
Agrega entrypoint para la tabla de temporadas y maxima cantidad de nidos

efren-cabrera committed:
Agrega código para desplegar lambdas en la tabla

efren-cabrera committed:
Agrega gitignore de archivos comunes para go y python

efren-cabrera committed:
Quita sección obsoleta del archivo de configuración de travis

efren-cabrera committed:
Agrega objetivo para construir la imagen

efren-cabrera committed:
Agrega dependencia a archivo de datos

efren-cabrera committed:
Agrega log a api de datos

efren-cabrera committed:
Agrega try para evitar que la api falle y regrese nan

efren-cabrera committed:
Cambia lugar donde se crea la tabla

efren-cabrera committed:
Agrega cabecera de la tabla

efren-cabrera committed:
Regresa solo valor de lambda

efren-cabrera committed:
Refactoriza sección que dibuja tabla

efren-cabrera committed:
Agrega tooltip a los polígonos

efren-cabrera committed:
Quita código innecesario y da formato al código

efren-cabrera committed:
Agrega descripción de lambda

efren-cabrera committed:
Cambia estilo de la tabla

efren-cabrera committed:
Agrega primer versión para dibujar gráfica

efren-cabrera committed:
Cambia posición de la barra lateral

efren-cabrera committed:
Cambia posición de control de zoom

efren-cabrera committed:
Quita código que mostrava hover y agrega etiquetas

efren-cabrera committed:
Corrige formato de ejes

efren-cabrera committed:
Corrige nombre de variables

efren-cabrera committed:
Merge remote-tracking branch 'github/feature/add_data' into feature/agrega_api_lambdas

efren-cabrera committed:
Trae cambios de la otra rama que no se hicieron de forma automática

devarops committed:
Corrige bugs

efren-cabrera committed:
Cambia dirección a donde se hace la petición

Con este cambio debería de poder funcionar en el servidor

devarops committed:
Cambia dirección a donde se hace la petición

devarops committed:
Merge pull request #3 from IslasGECI/update-solicitude-address

Cambia dirección a donde se hace la petición

devarops committed:
Limpia `.gitignore`

devarops committed:
Agrega Dockerfile

devarops committed:
Elimina diagonales innecesarias

nepito commented on api-lambdas/Dockerfile:
```diff
@@ -0,0 +1,5 @@
+FROM python:3
+COPY . /workdir/
+WORKDIR /workdir
+RUN pip install --requirement requirements.txt
```
```suggestion
RUN pip install flask \
         pandas \
         scipy
```

devarops replied:
OK. Pero pondré flask en su propia línea

nepito commented on api-lambdas/Dockerfile:
```diff
@@ -0,0 +1,5 @@
+COPY . /workdir/
+WORKDIR /workdir
+RUN pip install --requirement requirements.txt
+CMD ["python", "app.py"]
```
Supongo que si hay muchos `Dockerfile` también podrían haber algunos `Makefile` y ahí  poner `python app.py` como primer objetivo.

devarops replied:
OK

nepito commented on api-lambdas/requirements.txt:
```diff
@@ -0,0 +1,3 @@
+flask
```
Borraríamos este archivo si aceptas aquella sugerencia del `Dockerfile` de la carpeta `api-lambdas`.

devarops replied:
Listo

nepito commented on front/public_html/js/main.js:
```diff
@@ -0,0 +1,104 @@
+/**
```
¿Qué te parece cambiarle la terminación a este archivo? de `jc` a `tc`

devarops replied:
Te refieres de `js` a `ts`, cierto? Si es así muevo el archivo de `front/public_html/js/` a `front/src` y lo compilo en el `Dockerfile`.

devarops replied:
¿Te estabas imaginando algo así en el `Dockerfile`?
```
FROM node:lts AS compila_typescript
COPY . /workdir
WORKDIR /workdir
RUN npm install --global typescript
RUN tsc src/*.ts --outDir public_html/js

FROM httpd:2
COPY --from=compila_typescript /workdir/public_html/ /usr/local/apache2/htdocs/
```

evaristor replied:
@nepito?

efren-cabrera replied:
Si aún no hacen esto, estaría mejor hacer el cambio en otra rama

devarops replied:
Yo creo que no haremos esto todavía. Primero lo echaremos a andar nuevamente

nepito replied:
ok, eso suena muy bien

nepito reviewed (commented):
Creo que debo de dividir las futuras revisiones, tal vez solo un tipo de archivos o por _compose_.

devarops committed:
Mueve lista de requerimientos a Dockerfile

devarops commented:
@nepito:

> Creo que debo de dividir las futuras revisiones, tal vez solo un tipo de archivos o por compose.

No será fácil hacer manejable esta revisión. Yo creo con que limitemos cada tanda a 25 minutos es suficiente.

devarops committed:
Usa puertos sólo en el intervalo de 8000 a 8999

porque en el futuro quiero exponer sólo ese intervalo en el servidor islasgeci.org ya que actualmente se expone un intervalo demasiado amplio

devarops committed:
Elimina configuración de Travis ya que actualmente no está corriendo

podemos agregar GitHub Actions en el futuro o Travis nuevamente

devarops committed:
Agrega `Makefile` a la carpeta `api-lambdas`

nepito commented:
> @nepito:
> 
> > Creo que debo de dividir las futuras revisiones, tal vez solo un tipo de archivos o por compose.
> 
> No será fácil hacer manejable esta revisión. Yo creo con que limitemos cada tanda a 25 minutos es suficiente.

Hagamos que pasa (o truene) el actions

devarops commented:
> > @nepito:
> > > Creo que debo de dividir las futuras revisiones, tal vez solo un tipo de archivos o por compose.
> > 
> > 
> > No será fácil hacer manejable esta revisión. Yo creo con que limitemos cada tanda a 25 minutos es suficiente.
> 
> Hagamos que pasa (o truene) el actions

OK. Enonces agregare un archivo de configuración para el GitHub Actions

devarops committed:
Agrega archivo de configuración de GitHub Actions

devarops committed:
Mueve el _phony_ `run` al inicio para ejecutarlo por default con make

devarops committed:
Corrige nombre de Dockerfiles

efren-cabrera commented:
Parece que aquí truena por los datos, estará difícil que yo les ayude ahí, no sé como están manejando los datos actualmente. Me parece que antes los descargaba desde el servidor de los Notebooks 🤔. Pero sabíamos que eso era tmeporal

devarops commented:
@nepito Ya agregué la primer versión del archivo de configuración de GitHub Actions para que aparezca la tachita ❌

efren-cabrera commented:
Esta es la parte que falla, los jala directo de Bitbucket, pero no sé si ese repo aún exista en esa dirección

```
$(archivoDatos):
	mkdir --parents $(@D)
	curl --output $@ --user ${BITBUCKET_USERNAME}:${BITBUCKET_PASSWORD} https://bitbucket.org/IslasGECI/datos-binarios/raw/$(versionDatos)/$(@F)
```

devarops commented:
> Esta es la parte que falla, los jala directo de Bitbucket, pero no sé si ese repo aún exista en esa dirección

¡Excelente! Mañana le echaré un 👁️

devarops committed:
Ejecuta el Makefile de la raiz para conseguir los datos requeridos

devarops committed:
Define directorio de trabajo

devarops committed:
Usa doble signo de pesos para hacer referencia a variables de entorno

devarops committed:
Agraga credenciales de Bitbucket como variables de entorno

devarops committed:
Pasa variables de entorno al contenedor

devarops committed:
Descarga los datos en lugar de ejecutar Docker Compose

devarops committed:
Agrega archivos de configuración de GitHub Actions para develop y master

devarops commented:
Ya conseguí la ✔️ palomita 🙌

@nepito: Por favor me lo podrías aprobar?

nepito commented on .github/workflows/develop.yml:
```diff
@@ -0,0 +1,36 @@
+      env:
+        DOCKER_USERNAME: ${{ secrets.DOCKER_USERNAME }}
+        DOCKER_PASSWORD: ${{ secrets.DOCKER_PASSWORD }}
+      run: echo "${DOCKER_PASSWORD}" | docker login -u "${DOCKER_USERNAME}" --password-stdin
```
por favor usa la versión larga:
```suggestion
      run: echo "${DOCKER_PASSWORD}" | docker login --username "${DOCKER_USERNAME}" --password-stdin

```

nepito commented on .github/workflows/master.yml:
```diff
@@ -0,0 +1,36 @@
+      env:
+        DOCKER_USERNAME: ${{ secrets.DOCKER_USERNAME }}
+        DOCKER_PASSWORD: ${{ secrets.DOCKER_PASSWORD }}
+      run: echo "${DOCKER_PASSWORD}" | docker login -u "${DOCKER_USERNAME}" --password-stdin
```
Por favor utiliza la versión larga:
```suggestion
      run: echo "${DOCKER_PASSWORD}" | docker login --username "${DOCKER_USERNAME}" --password-stdin

```

nepito commented on Dockerfile:
```diff
@@ -1,29 +1,8 @@
+    curl \
+    docker.io \
+    make
+RUN curl -L "https://github.com/docker/compose/releases/download/1.24.1/docker-compose-$(uname -s)-$(uname -m)" -o /usr/local/bin/docker-compose && \
```
Por favor utiliza la versión larga:
```suggestion
RUN curl --location "https://github.com/docker/compose/releases/download/1.24.1/docker-compose-$(uname -s)-$(uname -m)" --output /usr/local/bin/docker-compose && \
```

nepito commented on Makefile:
```diff
@@ -1,17 +1,17 @@
+
+$(archivoDatos):
+	mkdir --parents $(@D)
+	curl --output $@ --user $${BITBUCKET_USERNAME}:$${BITBUCKET_PASSWORD} https://bitbucket.org/IslasGECI/datos-binarios/raw/$(versionDatos)/$(@F)
```
¿Por qué la variable `versionDatos` está entre paréntesis y no entre llaves?

efren-cabrera replied:
Entre paréntesis es porque es una variable del Makefile, aquí toma la parte del directorio que viene en la variable. Si mi memoria no me falla cuando usas llaves te refieres a una variable de entorno 🤔.

efren-cabrera replied:
Entre paréntesis es porque es una variable del Makefile, aquí toma la parte del directorio que viene en la variable. Si mi memoria no me falla cuando usas llaves te refieres a una variable de entorno 🤔.

devarops replied:
> To substitute a variable’s value, write a dollar sign followed by the name of the variable in parentheses or braces: either ‘$(foo)’ or ‘${foo}’ is a valid reference to the variable foo. This special significance of ‘$’ is why you must write ‘$$’ to have the effect of a single dollar sign in a file name or recipe.

- [Basics of Variable References](https://www.gnu.org/software/make/manual/html_node/Reference.html)

devarops replied:
Checa @nepito lo que dice este tal [Etan Reisner](https://stackoverflow.com/questions/25185607/whats-the-difference-between-parenthesis-and-curly-bracket-syntax-in-ma#comment39217556_25185742):

> I use the $() in make to avoid causing myself confusion (more than already exists) between make and shell variables.

nepito commented on api-datos/cmd/tablacruce/main.go:
```diff
@@ -0,0 +1,45 @@
+	"database/sql"
+	"log"
+	"net/http"
+
```
si no es necesario este espacio por favor quítalo

nepito commented on api-datos/Dockerfile:
```diff
@@ -0,0 +1,28 @@
+FROM golang:1.12.9 AS compila_go
+WORKDIR /workdir/cmd
+COPY ./cmd /workdir/cmd
+RUN go build -o ../bin/tablacruce ./tablacruce
```
No estoy seguro que acepte la versión larga:
```suggestion
RUN go build --output ../bin/tablacruce ./tablacruce
```

devarops replied:
Parece que aquí no hay versión larga

nepito commented on api-datos/cmd/tablacruce/main.go:
```diff
@@ -0,0 +1,45 @@
+			"%s - - \"%s %s\"",
+			requesterIP,
+			r.Method,
+			r.RequestURI,
```
¿Podrás quitar la coma?
```suggestion
			r.RequestURI
```

efren-cabrera replied:
No sé puede, eso es parte del lenguaje

nepito commented on docker-compose.yml:
```diff
@@ -0,0 +1,14 @@
+  front:
+    build: front
+    ports: 
+      - "8500:80"
```
En este archivo parece que está la información de los puertos

nepito commented on docker-compose.yml:
```diff
@@ -0,0 +1,14 @@
+  front:
+    build: front
+    ports: 
+      - "8500:80"
```
En este archivo parece que está la información de los puertos

efren-cabrera replied:
Solo tengan cuidado seguramente esto no son los puertos que deben cambiar para que funcione. Ahorita lo que truena es el front, no encuentra la api dónde esperaría, allá es donde habría que mover el puerto. Pero eso se debería hacer en otro _pull request_ este está grande y viejo.

nepito reviewed (changes requested):
Listo @evaristor aquí te van mis observaciones

nepito reviewed (changes requested):
Listo @evaristor aquí te van mis observaciones

devarops commented:
> Listo @evaristor aquí te van mis observaciones

Gracias, @nepito. Incorporo las correcciones en la primer oportunidad.

devarops committed:
Elimina coma innecesaria

devarops committed:
Elimina espacio innecesario

devarops committed:
Usa versión larga

devarops committed:
Usa versión larga de argumentos de cURL

devarops committed:
Uniformiza el uso de llaves y paréntesis al rededor de las variables

devarops committed:
Elimina comillas alrededor de las variables 

ya que con las comillas las variable no se expandían como esperaba y el GitHub Actions tronaba

devarops committed:
Agrega coma que había quitado

para tratar de resolver el siguiente error:

```
syntax error: unexpected newline, expecting comma or )
```

devarops committed:
Elimina la coma que quedaba volando

nepito commented on front/public_html/js/data.js:
```diff
@@ -17,7 +17,7 @@ function getIslands() {
 function getSeasons() {
     return new Promise((resolve, reject) => {
-        getDataFromEntryPoint("/temporadas", resolve);
+        getDataFromEntryPoint("http://islasgeci.org:8510/api-datos/temporadas", resolve);
```
Tiene finta de número mágico, al menos sería posible encapsular en una variable
```suggestion
        cons pathData = "http://islasgeci.org:8510/api-datos/temporadas";
        getDataFromEntryPoint(pathData, resolve);
```
Los cambios en este archivo muestran que no está de más crear una constante. Tal vez debería de ser: `let path_data = "http://islasgeci.org:8510/api-datos/temporadas";`

devarops replied:
@nepito: Atendí esto mediante unas variables globales al inicio del _script_. Sin pruebas, es difícil saber si esta refactorización funciona correctamente o no, así que me persigné antes de consignarlo y después de pusharlo. 🤞

nepito commented on front/public_html/js/main.js:
```diff
@@ -0,0 +1,104 @@
+    if (lambda < 0) {
+        return "red";
+    }
+    return "orange";
```
Qué será esto, un quinto caso?

efren-cabrera replied:
Seria cuando lbda sea 0.

devarops replied:
Pero la $\lambda$ no puede ser <= 0, no?

nepito commented on front/public_html/js/main.js:
```diff
@@ -0,0 +1,104 @@
+}
+
+function getLambdaColor(lambda) {
+    if (lambda === "nan")
```
Creo que lo podrías poner igual que los otros tres casos, con llaves
```suggestion
    if (lambda === "nan") {
```

nepito commented on front/public_html/js/main.js:
```diff
@@ -0,0 +1,104 @@
+
+function getLambdaColor(lambda) {
+    if (lambda === "nan")
+        return "white";
```
si aceptas el cambio de arriba, entonces este se vería algo así:
```suggestion
        return "white";
        }
```

nepito commented on front/public_html/js/draw.js:
```diff
@@ -0,0 +1,157 @@
+
+function drawTimeSerie(tablaHistorica) {
+    let margenes = { superior: 10, derecha: 30, inferior: 30, izquierda: 70 };
+    let ancho_grafica = 460 - margenes.izquierda - margenes.derecha;
```
tal vez podrías cambiar esto a _lowerCamelCase_ como en el resto del archivo

nepito reviewed (changes requested):
Listo @evaristor aquí te dejo mis comentarios

devarops commented:
> Listo @evaristor aquí te dejo mis comentarios

Muchas gracias, @nepito.

devarops committed:
Agrega llaves para que sea consistente con el resto de los casos

Co-Authored-By: nepito <braulio.rojas@islas.org.mx>

devarops committed:
Agrega comentario para describir caso

devarops committed:
Cambia snake_case a lowerCamelCase por consistencia

devarops committed:
Encapsula los URLs de los entry points en variables globales

devarops commented:
Listo, @nepito. Por favor le podrías echar un 👁️ nuevamente cuando tengas chance 🍅?

nepito commented on front/public_html/js/main.js:
```diff
@@ -0,0 +1,105 @@
+    if (lambda > 0 && lambda < 1) {
+        return "yellow";
+    }
+    if (lambda < 0) {
```
Hablé con el experto en crecimientos poblacionales y él me comenta que no tiene sentido una tasa de crecimiento poblacional negativa. Por lo menos en su forma: $N = N_o\lambda^t$.
Por lo que creo que deberías de cambiar  la función `getLambdaColor` si el modelo de que usamos es el de arriba

devarops committed:
Cambia el código de colores considerando un intervalo de 0 a 1

nepito commented on api-datos/cmd/tablacruce/modelos.go:
```diff
@@ -23,6 +23,11 @@ type Island struct {
 }
 
+// NewIsland Crea una nueva estructura de tipo Island
+func NewIsland(id int, nombre string, longitud, latitud float64) Island {
```
Parece que falta especificar el tipo del argumento de entrada:
```suggestion
func NewIsland(id int, nombre string, longitud float64, latitud float64) Island {
```

nepito commented on api-datos/cmd/tablacruce/modelos.go:
```diff
@@ -33,11 +38,15 @@ type CountTableRow struct {
-	return Island{id, nombre, longitud, latitud}
-}
-
 func NewCountTableRow(nombreEspecie string, nombreIsla string, maximoNidos int, temporada int, latitud, longitud float64, codigo string) CountTableRow {
```
Parece que falta agregar el tipo de la variable `latitud`:
```suggestion
func NewCountTableRow(nombreEspecie string, nombreIsla string, maximoNidos int, temporada int, latitud float64, longitud float64, codigo string) CountTableRow {
```

nepito commented on api-datos/cmd/tablacruce/modelos.go:
```diff
@@ -33,11 +38,15 @@ type CountTableRow struct {
+	MaximoNidos int
+}
+
+func NewHistoricTableRow(temporada, maximoNidos int) HistoricTableRow {
```
Parece que falta agregar el tipo del argumento `temporada`:
```suggestion
func NewHistoricTableRow(temporada int, maximoNidos int) HistoricTableRow {
```

nepito commented on api-lambdas/lambdas/get_lambda.py:
```diff
@@ -0,0 +1,13 @@
+import numpy as np
+
+
+def ley_potencia(T, Lambda, No):
```
Por favor usa el mismo idioma al nombrar las funciones a lo largo de este archivo
```suggestion
def power_law(T, Lambda, No):
```

nepito reviewed (changes requested):
Listo @evaristor aquí te dejo mis comentarios

devarops commented:
Muchas gracias @nepito

devarops committed:
Acepta sugerencias del @nepito

Co-authored-by: nepito <braulio.rojas@islas.org.mx>

devarops committed:
Cambia nombre de función a inglés para que estén en el mismo idioma

devarops commented:
Listo @nepito. Por favor podrías echarle un :eye: nuevamente?

nepito reviewed (approved):
Listo @evaristor .

devarops commented:
No te creo @nepito. :scream: Hasta no ver el _merge_ no te creeré.

devarops commented:
:exploding_head: :raised_hands:

devarops commented:
Gracias, @nepito
